### ISSUE TITLE
Audit learned graph checkpoint replay through ProgramBundle

### ISSUE THREAD

**Issue by Ahmadfauzi34** [Ahmadfauzi34/ai_agent_builder#175]
Follow-up to merged PR #174, which proved the end-to-end `EsOptimizer.ask -> graph parameter apply -> CompiledGraph.run -> objective -> EsOptimizer.tell` lifecycle and showed that public reusable binding/cache semantics are not currently justified.

This issue asks a narrower question before introducing any production controller or checkpoint API:

> Can the **existing** `ProgramBundle(include_state=true)` serve as the deterministic promotion/checkpoint artifact for a graph trained through the canonical graph-parameter binding?

## Research workload

Use a deterministic small regression graph and the existing strict ES path:

```text
fresh graph
  -> graphParameterLayout.total_len
  -> EsOptimizer.ask
  -> setGraphParametersFlat
  -> CompiledGraph.run over fixed dataset
  -> finite objective
  -> EsOptimizer.tell
  -> best candidate
  -> apply best
  -> exportProgramBundle(include_state=true)
  -> importProgramBundle into a fresh LayerRegistry
  -> replay graph / outputs / loss
```

## Required proofs

- training materially improves the fixed objective;
- optimizer dimension is derived only from canonical graph-parameter layout;
- best candidate is finite and reads back exactly after apply;
- mutable training preserves structural `program-identity.v1` and graph-parameter binding identity/layout;
- stateful ProgramBundle import into a fresh Registry reconstructs an equivalent CompiledGraph;
- imported graph preserves program identity and graph-parameter binding identity/layout;
- imported flat graph parameters equal the trained best candidate exactly;
- imported outputs/loss reproduce the trained graph within a strict deterministic tolerance;
- malformed/truncated learned bundle import fails closed and leaves an already-populated target Registry unchanged;
- repeated stateful export determinism is measured and reported; do not invent a new identity protocol merely to force byte equality.

A host SHA-256 of exported bundle bytes may be recorded as evidence only. It is **not** a new protocol identity or authorization token.

## Decision boundary

If the existing stateful ProgramBundle cleanly replays the learned graph, classify the checkpoint path as **KEEP** and defer any separate checkpoint/state-identity surface until a concrete missing workload proves one is needed.

If it fails, isolate the exact missing semantic boundary before proposing production code.

## Non-goals

- no production training/controller facade;
- no reusable/public graph-parameter cache;
- no new checkpoint schema or graph-state identity;
- no ES formula/state changes;
- no graph-plan changes;
- no Math Program v10 or other math semantics;
- no autodiff/backprop;
- no WASM public-surface change;
- no Resolution/Authorization/proof/execution authority coupling.

**Comment by Ahmadfauzi34** [comment]
Research conclusion from PR #176 and exact packaged-artifact verification:

**Decision: KEEP the existing `ProgramBundle(include_state=true)` as the learned graph checkpoint/replay artifact.**

Evidence:
- full PR CI green;
- packaged checkpoint gate PASS;
- independent no-rebuild audit against the exact workflow artifact PASS;
- learned flat parameters, outputs, loss, program identity, graph-parameter binding identity, and layout replay correctly after import into a fresh Registry;
- malformed/truncated learned bundle import is atomic/fail-closed against a populated target Registry.

Important nuance discovered by the second-opinion audit: exact ProgramBundle bytes/SHA are stable for repeated export of the **same loaded state** and for import/re-export of that representation, but are not canonical across fresh processes because Burn state serialization carries run/process-specific tensor identifier metadata. Numeric parameters and semantic replay remain stable.

So the correct boundary is semantic checkpoint/replay, not cross-process byte identity. Do not introduce a new state identity, checkpoint schema, or authorization/proof coupling merely to canonicalize Burn serialization. If a future workload truly requires a portable canonical `I_state`, that should be a separate protocol with its own proof, not a reinterpretation of ProgramBundle SHA.
